### DISCUSSION TITLE
anyone know what "box" and "boxes" results represent?

### DISCUSSION THREAD

**Discussion by ericblade** [ericblade/quagga2#320] [Q&A]
... I'm slightly confused as to what the "box" and "boxes" results represent.  'box' appears to be a 4 length array of 2-length arrays, but i'd expect a box would just be a single point pair.. rather than 4 of them.  'boxes' just seems to be an array of that array, though that might be different if multiples are encountered.

**Comment by Denacke** [comment]
If I were to guess without delving in it deeper at the moment, i'd figure those are the 4 x&y-coordinates of the corners of the box? 

Thank you for maintaining this repo by the way!

**Comment by ericblade** [comment]
yeah, i think it might be a good idea to write a sample that visualizes the results in an obvious way (maybe draw some circles/dots at each point in the 'box' / 'boxes' arrays) to get a better idea what is represented there.  

Since you can create a box using just 2 points, one corner, and the opposite corner, it slightly surprised me that it appears to have 4 points, but I guess that makes it easier without having to math out the other points.  I guess I'm used to game engines and UI building engines, where you can easily supply a box with just two points. :-D

So, you're probably right, but I think I might try to add some additional things to the react sample that indicate what the various outputs represent specifically . . . since I'm not the one who designed the output, and it isn't all entirely clear what all the things represent, or their purpose.

Also, you're welcome!

**Comment by ChristophKaser** [comment]
I might be wrong, but I don't think 2 points are enough to fully specify a box unless you require it to be axis-aligned. Since quagga supports rotated boxes, this might be the reason for 4 coordinates (just a guess, I did not look at the source)

**Comment by ericblade** [comment]
I hadn't considered that, you are probably right there.  *thinks* have i ever tried to draw a box that wasn't axis aligned to the screen? i am not sure.

**Comment by TonyKhorouzan** [comment]
My understanding is that the four points are a set of points defining the corners of a quadrilateral.  The region is not necessarily oriented along the x and y axis.   

Initially, the quadrilateral starts off with a clockwise ordering from the top-left corner and is aligned with the X and Y axis and may actually be a rectangle.  The initial array is indexed:
            0 = Top Left
            1 = Top Right
            2 = Bottom Right
            3 = Bottom Left
Their are aggregate operations (BoxFromPatches) and rotations, so this relationship is not held, and resulting "boxes" can be weird quadrilaterals,  Also, it then becomes ambiguous which corner is the top-left corner.  Consider a square tilted 45 degrees.  One point can be top-most and another can be left-most, but neither, both, or either can be considered top-left.

**Comment by pogorv12** [comment]
> ... I'm slightly confused as to what the "box" and "boxes" results represent. 'box' appears to be a 4 length array of 2-length arrays, but i'd expect a box would just be a single point pair.. rather than 4 of them. 'boxes' just seems to be an array of that array, though that might be different if multiples are encountered.

Hello!

Box - is the decoded area that provided the result. Boxes - is the list of recognized areas that potentially located barcodes. 

So. If you have one barcode on an image you will have on element in Boxes. If you have plenty barcodes, than you will have a few boxes, and the first decoded will be returned as the box.
